cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -252,6 +280,30 @@ func (q *Queue) SetIndexed(opts IndexOptions, state indexState) {
 	item.indexState = state
 	if state != indexStateFail {
 		item.indexed = reflect.DeepEqual(opts, item.opts)
+		item.consecutiveFailures = 0
+		item.backoffUntil = time.Unix(0, 0)
+	} else {
+		backoffDuration := time.Duration(item.consecutiveFailures+1) * q.backoffDuration
+
+		if backoffDuration > q.maxBackoff {
+			backoffDuration = q.maxBackoff
+		} else {
+			item.consecutiveFailures++
+		}
+		item.backoffUntil = time.Now().Add(backoffDuration)
+
+		q.logger.Debug("Backoff subsequent attempts to index repository",
+			sglog.String("repo", item.opts.Name),
+			sglog.Uint32("id", item.repoID),
+			sglog.Duration("backoff_duration", backoffDuration),
+			sglog.Time("backoff_until", item.backoffUntil),
+		)
+
+		// if we backoff indexing for a repo then it should never pop for processQueue
```

Comment:
`SetIndexed` is only called in `processQueue()` where item was already popped and therefore heapIdx should be -1, so we probably don't need this, but this additional logic is just precautionary.

Reply:
this is not correct unfortunetly. We don't hold the lock while indexing so concurrently we can re-add the item to the queue.

Reply:
I see, yeah I forgot the bump() or iterateIndexOptions() from the sync loop could add during our processQueue.